File: CSharpCodeBase/components/controlsystem/actions/actionmove.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class actionmove :action{
 public void init (self, entity, target);
       Action.init(self, 1, true, "move");
       this.isKeyboardControl = true;
       this.currentInput = Vector3();
       this.entity = entity;
       this.target = target;
 }})
 public void SwitchToKeyboard(){
   this.isKeyboardControl = true;
 }
 public void SwitchToMouse(){
   this.isKeyboardControl = false;
 }
 public void OnStartRunning(){
   if(this.target != null  ){ 
     self:SwitchToMouse();
     this.entity.mover:SetGoal(this.target);
   }
   this.entity.mover:ResetSpeed();
 }
 public void Update(){
   var actionResult = true;
   if(GameController.inputService:LeftStickYIsPressed()  ||   GameController.inputService:LeftStickXIsPressed()  ){ 
     this.currentInput = GameController.inputService:LeftStickValues();
     this.currentInput:RotateAroundY(GameController.camera.angle) ;
     this.entity.mover:SetInput(this.currentInput);
     actionResult = false;
   }
   if(this.isKeyboardControl  ){ 
     if(Input.GetMouseButtonDown(1)  ){ 
       this.entity.mover:SetGoal(Input.RaycastMouseOnTerrain());
       self:SwitchToMouse();
       actionResult = false;
     }
     
     if(GameController.inputService:GetMouseButton(1)  &&  this.currentInput:Length() == 0  ){ 
       self:SwitchToMouse();
       actionResult = false;
     }
   }
   
   if(not this.isKeyboardControl  ){ 
     if(GameController.inputService:GetMouseButton(1)  ){         
       this.entity.mover:SetGoal(Input.RaycastTargetEntity()  ||  Input.RaycastMouseOnTerrain());
       actionResult = false;
     }    
     
     if(not this.entity.mover:IsHaveGoal()  ||  GameController.inputService:LeftStickYIsPressed()  ||  GameController.inputService:LeftStickXIsPressed()  ){ 
       self:SwitchToKeyboard();
     }
     
     if(this.entity.mover:IsHaveGoal()  ){ 
       actionResult = false;
     }
   }
   return actionResult;
 }
 public void OnStopRunning(){
   this.entity.mover:Stop();
 }
 }}

File: CSharpCodeBase/components/controlsystem/actions/actionmove.lua
local Action = require 'components.controlsystem.action'
local Class = require 'utils.hump.class'
local Vector3 = require 'utils.hump.vector3'

local ActionMove = Class({__includes=Action, init = function (self, entity, target)
      Action.init(self, 1, true, 'move')
      self.isKeyboardControl = true
      self.currentInput = Vector3()
      self.entity = entity
      self.target = target
end})

function ActionMove:SwitchToKeyboard()
  self.isKeyboardControl = true
end

function ActionMove:SwitchToMouse()
  self.isKeyboardControl = false
end

function ActionMove:OnStartRunning()
  if self.target ~= nil then
    self:SwitchToMouse()
    self.entity.mover:SetGoal(self.target)
  end
  self.entity.mover:ResetSpeed()
end

function ActionMove:Update()
  local actionResult = true
  if GameController.inputService:LeftStickYIsPressed() or  GameController.inputService:LeftStickXIsPressed() then
    self.currentInput = GameController.inputService:LeftStickValues()
    self.currentInput:RotateAroundY(GameController.camera.angle) 
    self.entity.mover:SetInput(self.currentInput)
    actionResult = false
  end
  if self.isKeyboardControl then
    if Input.GetMouseButtonDown(1) then
      self.entity.mover:SetGoal(Input.RaycastMouseOnTerrain())
      self:SwitchToMouse()
      actionResult = false
    end
    
    if GameController.inputService:GetMouseButton(1) and self.currentInput:Length() == 0 then
      self:SwitchToMouse()
      actionResult = false
    end
  end
  
  if not self.isKeyboardControl then
    if GameController.inputService:GetMouseButton(1) then        
      self.entity.mover:SetGoal(Input.RaycastTargetEntity() or Input.RaycastMouseOnTerrain())
      actionResult = false
    end    
    
    if not self.entity.mover:IsHaveGoal() or GameController.inputService:LeftStickYIsPressed() or GameController.inputService:LeftStickXIsPressed() then
      self:SwitchToKeyboard()
    end
    
    if self.entity.mover:IsHaveGoal() then
      actionResult = false
    end
  end
  return actionResult
end

function ActionMove:OnStopRunning()
  self.entity.mover:Stop()
end
return ActionMove